### PR TITLE
Add endpoint to connection string

### PR DESCRIPTION
**Pull request by vadim2404**
## Describe your changes
We should force our users to use endpoint wording vs. project.

## Issue ticket number and link
Closes neondatabase/cloud#4928

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Review by hlinnaka — APPROVED** [review]
There's an accidental change to `vendor/postgres-v15`. Other than that, +1

**Comment by vadim2404** [comment]
> There's an accidental change to `vendor/postgres-v15`. Other than that, +1

Yes, thanks. I didn't properly updated submodules. Thanks for highlighting

**Comment by vadim2404** [comment]
@SergeyMelnikov, added in these places too. Rely on @funbringer's review

**Review by funbringer — COMMENTED** [review]
Looks good overall. I'd change a few things:

* First, extract stuff like this into a function:

```
.starts_with(b"project=") {
            // project is deprecated, but we need to have it for backward compatibility.
            bytes.strip_prefix(b"project=")
        } else {
            bytes.strip_prefix(b"endpoint=")
        }
````

E.g. `fn parse_endpoint_param(s: &str) -> Option<&str>;`.

* Add `endpoint=` variant to `test_runner/regress/test_proxy.py` tests, probably using pytest's fixtures.

**Comment by vadim2404** [comment]
> Looks good overall. I'd change a few things:
> 
> * First, extract stuff like this into a function:
> 
> ```
> .starts_with(b"project=") {
>             // project is deprecated, but we need to have it for backward compatibility.
>             bytes.strip_prefix(b"project=")
>         } else {
>             bytes.strip_prefix(b"endpoint=")
>         }
> ```
> 
> E.g. `fn parse_endpoint_param(s: &str) -> Option<&str>;`.
> 
> * Add `endpoint=` variant to `test_runner/regress/test_proxy.py` tests, probably using pytest's fixtures.

Addressed your comments.

File: proxy/src/auth/credentials.rs [hunk 1]
```diff
@@ -61,7 +61,15 @@ impl<'a> ClientCredentials<'a> {
         // Project name might be passed via PG's command-line options.
         let project_option = params
```
Comment (bayandin) [inline proxy/src/auth/credentials.rs:62]: Do we want to rename it to `endpoint_option`?

File: proxy/src/auth/credentials.rs [hunk 2]
```diff
@@ -61,7 +61,15 @@ impl<'a> ClientCredentials<'a> {
...
-            .and_then(|mut options| options.find_map(|opt| opt.strip_prefix("project=")))
+            .and_then(|mut options| {
+                options.find_map(|opt| {
+                    if opt.starts_with("project=") {
+                        opt.strip_prefix("project=")
+                    } else {
+                        opt.strip_prefix("endpoint=")
+                    }
+                })
+            })
```
Comment (bayandin) [inline proxy/src/auth/credentials.rs:72]: What will happen if someone passes both of these parameters in the options?

File: proxy/src/auth/credentials.rs (hunk 1 above)
Comment (vadim2404) [reply proxy/src/auth/credentials.rs:62]: not yet, I intentionally didn't changed the whole code base

File: proxy/src/auth/credentials.rs (hunk 2 above)
Comment (bayandin) [reply proxy/src/auth/credentials.rs:72]: > Do you think we need to abrupt the connection if two options exist?

I think it should use `endpoint` parameter if it exists, if not, then check `project` as a fallback.
Now it looks like some undefined behaviour, I'd like to have it consistent (an option with abrupting connection suits too)

File: test_runner/fixtures/neon_fixtures.py [hunk 3]
```diff
@@ -2237,6 +2237,19 @@ def static_proxy(
...
+    scope="function",
+    params=[
+        "project",
+        "endpoint",
+    ],
+)
+def proxy_endpoint_option_name(request: FixtureRequest) -> str:
+    assert hasattr(request, "param")
+    assert isinstance(request.param, str)
+    return request.param
```
Comment (funbringer) [inline test_runner/fixtures/neon_fixtures.py:2250]: To me putting this small piece in the fixtures library doesn't make much sense. It's very ad-hoc and hard to discover.

 I'd solve this using `pytest.mark.parametrize`:

```python
@pytest.mark.parametrize("name", ["endpoint", "project"])
def test_fizz_buzz(name: str):
    pass
```

File: proxy/src/auth/credentials.rs [hunk 4]
```diff
@@ -61,7 +61,21 @@ impl<'a> ClientCredentials<'a> {
...
+                // We support both `project` (deprecated) and `endpoint` options for backward compatibility.
+                // However, if both are present, we don't exactly know which one to use.
+                // Therefore we require that only one of them is present.
+                options
+                    .filter_map(|opt| {
+                        if opt.starts_with("project=") {
+                            opt.strip_prefix("project=")
+                        } else {
+                            opt.strip_prefix("endpoint=")
+                        }
```
Comment (funbringer) [inline proxy/src/auth/credentials.rs:74]: it's `parse_endpoint_param`

File: proxy/src/compute.rs [hunk 5]
```diff
@@ -280,6 +280,7 @@ fn filtered_options(params: &StartupMessageParams) -> Option<String> {
     let options: String = params
         .options_raw()?
         .filter(|opt| !opt.starts_with("project="))
+        .filter(|opt| !opt.starts_with("endpoint="))
```
Comment (funbringer) [inline proxy/src/compute.rs:283]: `filter(|opt| parse_endpoint_param(opt).is_none())`

File: proxy/src/auth/password_hack.rs [hunk 6]
```diff
@@ -6,18 +6,27 @@
...
+    }
+}
+
+fn parse_endpoint_param(bytes: &[u8]) -> Option<&[u8]> {
+    if bytes.starts_with(b"project=") {
+        // project is deprecated, but we need to have it for backward compatibility.
+        bytes.strip_prefix(b"project=")
+    } else {
+        bytes.strip_prefix(b"endpoint=")
     }
```
Comment (funbringer) [inline proxy/src/auth/password_hack.rs:30]:
```rust
s.strip_prefix(b"endpoint=").or_else(|| s.strip_prefix(b"project="))
```

File: test_runner/fixtures/neon_fixtures.py (hunk 3 above)
Comment (vadim2404) [reply test_runner/fixtures/neon_fixtures.py:2250]: It was my initial implementation, btw. But I followed your idea from initial comment

Comment (funbringer) [reply test_runner/fixtures/neon_fixtures.py]: Sorry if I gave a wrong impression; this is what I had in mind at that time.

**Review by funbringer — APPROVED** [review]

**Pull request merged**